remsfal-service/src/main/java/de/remsfal/service/boundary/authentication/SessionManager.java

```diff
@@ -1,152 +1,220 @@
+        return generateAccessToken(SessionInfo.builder().from(userAuth).build());
+    }
+
+    @Transactional
+    public NewCookie generateRefreshToken(String userId, String userEmail) {
+        String refreshToken = UUID.randomUUID().toString();
+        Optional<UserAuthenticationEntity> existingAuth = userAuthRepository.findByUserId(userId);
+        if (existingAuth.isEmpty()) {
+            UserAuthenticationEntity userAuthenticationEntity = new UserAuthenticationEntity();
+            userRepository.findByIdOptional(userId).ifPresentOrElse(userAuthenticationEntity::setUser, () -> {
+                throw new UnauthorizedException("User not found: " + userId);
+            });
+            userAuthenticationEntity.setRefreshToken(refreshToken);
+            userAuthRepository.persist(userAuthenticationEntity);
         } else {
-            // Generate symmetric 128 bit AES key
-            KeyGenerator keyGen = KeyGenerator.getInstance(KEY_ALGORITHM);
-            keyGen.init(128);
-            secretKey = keyGen.generateKey();
+            userAuthRepository.updateRefreshToken(userId, refreshToken);
         }
+        String jwt = jwtManager.createJWT(
+            SessionInfo.builder().userId(userId).userEmail(userEmail).claim("refreshToken", refreshToken)
+                .expireAfter(refreshTokenTimeout).build());
+        return buildCookie(REFRESH_COOKIE_NAME, jwt, (int) refreshTokenTimeout.getSeconds(), true);
     }
 
-    public String encryptSessionObject(final SessionInfo sessionInfo) {
-        // Create the header and payload
-        JWEHeader header = new JWEHeader(JWEAlgorithm.DIR, EncryptionMethod.A128GCM);
-        
-        // Create the JWE object and encrypt it
-        JWEObject jweObject = new JWEObject(header, sessionInfo.toPayload());
-        try {
-            jweObject.encrypt(new DirectEncrypter(secretKey));
-        } catch (JOSEException e) {
-            throw new InternalServerErrorException("Unable to encrypt session object", e);
+
+    public TokenRenewalResponse renewTokens(Map<String, Cookie> cookies) {
+        Cookie refreshCookie = cookies.get(REFRESH_COOKIE_NAME);
+
+        if (refreshCookie == null) {
+            throw new UnauthorizedException("No refresh token provided.");
         }
 
-        // Serialize to a compact JOSE
-        return jweObject.serialize();
+        SessionInfo refreshToken = decryptRefreshTokenCookie(refreshCookie);
+        UserAuthenticationEntity userAuth = checkValidRefreshToken(refreshToken);
+
+        return new TokenRenewalResponse(generateAccessToken(userAuth),
+            generateRefreshToken(refreshToken.getUserId(), refreshToken.getUserEmail()));
+
+    }
+
+    public SessionInfo decryptAccessTokenCookie(Cookie cookie) {
+        return jwtManager.verifyJWT(cookie.getValue());
     }
 
-    public SessionInfo decryptSessionObject(final String sessionObject) {
+    public SessionInfo checkValidUserSession(Map<String, Cookie> cookies) {
```

Comment:
Methods like checkValidUserSession, generateAccessToken, and generateRefreshToken could benefit from Javadoc to clarify their purpose and expected behavior, especially considering the complex logic they contain.